### PR TITLE
Device trust terraform provider

### PR DESCRIPTION
**Pull request by sshahcodes**
Terraform support for device trust. 

### Device type definition
Currently, there's two Device types implemented for Device Trust: 
1. [Device](https://github.com/gravitational/teleport/blob/master/api/proto/teleport/devicetrust/v1/device.proto) proto, which is the main definition used by all Device Trust related services.
2. [DeviceV1](https://github.com/gravitational/teleport/blob/master/api/proto/teleport/legacy/types/device.proto) used by this provider.

### Provider implementation
Unlike other existing resources, [DeviceV1](https://github.com/gravitational/teleport/blob/master/api/proto/teleport/legacy/types/device.proto) does not fully utilize the metadata fields such as `label`, `description`, etc. Only `metadata.name` is supported, specifically to be compatible with schema generation for terraform provider. Additionally, the [DeviceV1](https://github.com/gravitational/teleport/blob/master/api/proto/teleport/legacy/types/device.proto) embeds the `ResourceHeader` field. [protoc-gen-terraform](https://github.com/gravitational/protoc-gen-terraform/pull/29) is updated to support such embedded fields.
 
This PR adds a new fields to the template payload:
- `UUIDMetadataName` field: used to generate `Metadata.Name` as uuid value. The [`plural_resource.go.tpl`](https://github.com/gravitational/teleport-plugins/blob/a76be035e7498090958e05f356426d2a4084281d/terraform/gen/plural_resource.go.tpl) template is updated to accommodate `UUIDMetadataName` field. 

### Supported device trust CRUD fields
As for supported fields, only `asset_tag`, `os_type`, and `enroll_status` are supported by this provider. 
1. Create: Create device: `asset_tag`, `os_type`, `enroll_status` (optional as this value can be updated outside of Terraform and it won't be a good experience for our users to keep the `.tf` files in sync with the state )
4. Update: Update `enroll_status`: `enrolled -> not_enrolled`
5. Delete: Delete device.

Closes: https://github.com/gravitational/teleport.e/issues/939
Depends on https://github.com/gravitational/protoc-gen-terraform/pull/29

### CODE REVIEW

**Comment by sshahcodes** [comment]
Note to reviewer on failing CI test: This PR requires latest teleport version (master branch). What would be the way to update this before `v13` is released?

**Comment by marcoandredinis** [comment]
> Note to reviewer on failing CI test: This PR requires latest teleport version (master branch). What would be the way to update this before `v13` is released?

If you want the CI to pass, you must update the teleport dep in go.mod to point to current master's commit (or at least to a commit where your changes were made)
However, we can't merge that before the v13 release

**Comment by marcoandredinis** [comment]
Yeah, we should leave this PR until v13 gets released and during the non-production ready releases (alpha, beta, rc, ...) we should be able to update the API and the binary used to run tests

The binary used to run the tests is under the `TELEPORT_GET_VERSION` variable
It can be updated using `TELEPORT_GET_VERSION=v13.0.0 make update-test-version` (when that is released)

File: go.mod [hunk 1]
```diff
@@ -129,6 +127,6 @@ require (
 
 replace (
 	github.com/gogo/protobuf => github.com/gravitational/protobuf v1.3.2-0.20201123192827-2b9fcfaffcbf
-	github.com/gravitational/teleport/api => github.com/gravitational/teleport/api v0.0.0-20230405020520-6b7a47352ee7 // ref: tags/v12.2.1
+	github.com/gravitational/teleport/api => github.com/gravitational/teleport/api v0.0.0-20230404165006-e0796b6f1b74 // tag v12.1.0
```
Comment (nklaassen) [inline go.mod:130]: should this be updated to a 13 pre-release?

File: terraform/_gen/plural_resource.go.tpl [hunk 2]
```diff
@@ -286,7 +297,12 @@ func (r resourceTeleport{{.Name}}) Update(ctx context.Context, req tfsdk.UpdateR
 			resp.Diagnostics.Append(diagFromWrappedErr("Error reading {{.Name}}", err, "{{.Kind}}"))
 			return
 		}
-		{{if .IsPlainStruct -}}
+		{{if .MetadataNameAsID -}}
+		if {{.VarName}}Before.Metadata.Name != {{.VarName}}.Metadata.Name || {{.HasStaticID}} {
+			break
+		}
+        {{else -}}
+        {{if .IsPlainStruct -}}
```
Comment (nklaassen) [inline terraform/_gen/plural_resource.go.tpl:305]: these are spaces, the rest of the file is tabs

File: terraform/_gen/plural_resource.go.tpl [hunk 3]
```diff
@@ -295,6 +311,7 @@ func (r resourceTeleport{{.Name}}) Update(ctx context.Context, req tfsdk.UpdateR
 			break
 		}
 		{{- end}}
+        {{- end}}
```
Comment (nklaassen) [inline terraform/_gen/plural_resource.go.tpl:314]: these are spaces, the rest of the file is tabs

File: terraform/_gen/plural_resource.go.tpl [hunk 4]
```diff
@@ -191,7 +191,7 @@ func (r resourceTeleport{{.Name}}) Read(ctx context.Context, req tfsdk.ReadResou
 	}
 
 	var id types.String
-	diags = req.State.GetAttribute(ctx, path.Root("metadata").AtName("name"), &id)
+	diags = req.State.GetAttribute(ctx, {{if .MetadataNameAsID -}}path.Root("id"){{else -}}path.Root("metadata").AtName("name"){{end}}, &id)
```
Comment (nklaassen) [inline terraform/_gen/plural_resource.go.tpl:194]: Am I reading this right? You're reading into the `id` variable, and `if .MetadataNameAsID` you get `root.id`, else you get `root.metadata.name`? It seems backward

File: terraform/_gen/main.go [hunk 5]
```diff
@@ -73,6 +73,13 @@ type payload struct {
 	// IsPlainStruct states whether the resource type used by the API methods
 	// for this resource is a plain struct, rather than an interface.
 	IsPlainStruct bool
+	// ReadReturnValue indicates to read return value from Teleport Create/Update method.
+	// Reading return value is required in situations where data returned by Teleport is required
+	// to further proceed with Create/Update method in the provider.
+	ReadReturnValue bool
+	// MetadataNameAsID indicates to assign Metadata Value to an injected id field and vice versa.
+	// Currently used by device trust provider to assign Metadata.Name to injected id field.
```
Comment (nklaassen) [inline terraform/_gen/main.go:81]: could you maybe explain, do we use metadata.name from the terraform resource as the id in the teleport resource, or the other way? which part injects the id where?

**Review by nklaassen — COMMENTED** [review]
I'm curious what the expected use case for this is? Will admins be able to define all resource asset tags in terraform and then end users will be able to enroll their device with tsh, or something like that?

File: terraform/_gen/plural_resource.go.tpl (hunk 4 above)
Comment (sshahcodes) [reply terraform/_gen/plural_resource.go.tpl]: This field is no longer required as embedded fields are supported now (see https://github.com/gravitational/protoc-gen-terraform/pull/29). I have updated this PR.

**Review by mdwn — APPROVED** [review]

File: terraform/test/fixtures/device_trust_0_create.tf [hunk 6]
```diff
@@ -0,0 +1,8 @@
+
```
Comment (marcoandredinis) [inline terraform/test/fixtures/device_trust_0_create.tf:1]: 💅 Extra empty line

File: terraform/test/fixtures/device_trust_1_update.tf [hunk 7]
```diff
@@ -0,0 +1,16 @@
+
```
Comment (marcoandredinis) [inline terraform/test/fixtures/device_trust_1_update.tf:1]: 💅 Extra empty line

**Comment by marcoandredinis** [comment]
We need this issue fixed for the tests to pass
https://github.com/gravitational/teleport-plugins/issues/820

File: terraform/protoc-gen-terraform-devicetrust.yaml
[the remark this answers is not in the document]
Comment (sshahcodes) [reply terraform/protoc-gen-terraform-devicetrust.yaml:1]: Hmm, I think right. This work started with the [Device](https://github.com/gravitational/teleport/blob/master/api/proto/teleport/devicetrust/v1/device.proto#L31) proto but later switched to another proto added to legacy proto type. I never thought about adding this to `protoc-gen-terraform-teleport.yaml` afterward. Let me fix that.

File: terraform/test/device_trust_test.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+import (
+	"fmt"
+
+	"github.com/gravitational/trace"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+)
+
+func (s *TerraformSuite) TestDeviceTrust() {
```
[the remark this answers is not in the document]
Comment (sshahcodes) [reply terraform/test/device_trust_test.go]: added in https://github.com/gravitational/teleport-plugins/pull/801/commits/54c212430c9c029691b4ca2f345ed8dcb4dfedd7

File: terraform/protoc-gen-terraform-devicetrust.yaml
[the remark this answers is not in the document]
Comment (marcoandredinis) [reply terraform/protoc-gen-terraform-devicetrust.yaml:1]: Yes, you are correct 👍 
Sorry for the confusion

**Review by marcoandredinis — APPROVED** [review]
Can you please re-run the the `terraform-gen-tfschema` target?
I got some differences

Btw, after this gets merged and released (13.0.4 ?) we should update the Terraform Reference docs
Something like this:
https://github.com/gravitational/teleport/pull/26502 (I have to fix a "typo" before I can merge that one)

**Comment by sshahcodes** [comment]
This is scheduled for `13.1` but noted to update reference docs. Thanks

**Review by r0mant — APPROVED** [review]
Bot.

**Pull request merged**